### PR TITLE
dev: Enable IPv6 system setting for devcontainer environment.

### LINKED ISSUE #31263: `devcontainer` cilium installation fails with `could not load module ip6_tables`
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

https://docs.cilium.io/en/stable/contributing/development/dev_setup/#kind-based-setup-preferred

1. start devcontainer with vscode.
2. `make kind`, `make kind-image` and `make kind-install-cilium`
3. cilium pods failed to start, and never gets to be `RUNNING` state.

```bash
root@705e87b2a7e2:/go/src/github.com/cilium/cilium# kubectl logs -n kube-system cilium-q2cqp | grep fatal
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
level=fatal msg="failed to start: IPv6 is enabled and ip6tables modules initialization failed: could not load module ip6_tables: exit status 1 (try disabling IPv6 in Cilium or loading ip6_tables, ip6table_mangle, ip6table_raw and ip6table_filter kernel modules)" subsys=daemon
```

### Cilium Version

https://github.com/cilium/cilium/commit/02d04b34005c45d955a6850d3faf561d9eae8f78

### Kernel Version

Linux tomoyafujita 5.15.0-97-generic #107~20.04.1-Ubuntu SMP Fri Feb 9 14:20:11 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

kubeadm version: &version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.7", GitCommit:"84e1fc493a47446df2e155e70fca768d2653a398", GitTreeState:"clean", BuildDate:"2023-07-19T12:22:13Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

```shell
root@705e87b2a7e2:/go/src/github.com/cilium/cilium# kubectl logs -n kube-system cilium-q2cqp | grep fatal
Defaulted container "cilium-agent" out of: cilium-agent, config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)
level=fatal msg="failed to start: IPv6 is enabled and ip6tables modules initialization failed: could not load module ip6_tables: exit status 1 (try disabling IPv6 in Cilium or loading ip6_tables, ip6table_mangle, ip6table_raw and ip6table_filter kernel modules)" subsys=daemon
```

### Anything else?

disabling `ipv6.enabled` can resolve this error.

```yaml
diff --git a/contrib/testing/kind-common.yaml b/contrib/testing/kind-common.yaml
index 1be34fdebd..aaabca62a3 100644
--- a/contrib/testing/kind-common.yaml
+++ b/contrib/testing/kind-common.yaml
@@ -7,7 +7,7 @@ operator:
 ipam:
   mode: kubernetes
 ipv6:
-  enabled: true
+  enabled: false
 ipv4:
   enabled: true
 bpf:
```

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by fujitatomoya**
Fixes: #31263

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

Fixes: #31263

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -6,5 +6,13 @@
   "features": {
     "ghcr.io/devcontainers/features/docker-in-docker": {}
   },
+  "mounts": [
+    // To enable kernel modules for devcontainer
+    "source=/lib/modules,target=/lib/modules,type=bind,readonly",
+    "source=/usr/lib/modules,target=/usr/lib/modules,type=bind,readonly"
+  ],
+  "runArgs": [
+    "--sysctl", "net.ipv6.conf.all.disable_ipv6=0"
```
Comment (fujitatomoya) [inline .devcontainer/devcontainer.json:15]: without this, there will be system config conflict for IPv6. that is going to lead to the following error.

```bash
level=warning msg="Waiting for k8s node information" error="required IPv6 PodCIDR not available" subsys=daemon
```

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -6,5 +6,13 @@
   "features": {
     "ghcr.io/devcontainers/features/docker-in-docker": {}
   },
+  "mounts": [
+    // To enable kernel modules for devcontainer
+    "source=/lib/modules,target=/lib/modules,type=bind,readonly",
+    "source=/usr/lib/modules,target=/usr/lib/modules,type=bind,readonly"
```
Comment (fujitatomoya) [inline .devcontainer/devcontainer.json:12]: cilium also checks debug loadable module, i am not sure if we need it here? any thoughts?

**Comment by aditighag** [comment]
I don't use .devcontainer, so deferring to @sayboras's review.

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (sayboras) [reply .devcontainer/devcontainer.json:12]: I don't think it's a strict requirement

**Review by sayboras — APPROVED** [review]
Thanks a lot for your contribution, I can run cilium inside devcontainer without any error from this PR.

Edit: We can improve the experience further by pre-install required tools (e.g. kind, kubectl, and cilium-cli), however, this can be done in separate PR.

```
root@b67d3bb54a76:/go/src/github.com/cilium/cilium# cilium status
    /¯¯\
 /¯¯\__/¯¯\    Cilium:             OK
 \__/¯¯\__/    Operator:           OK
 /¯¯\__/¯¯\    Envoy DaemonSet:    OK
 \__/¯¯\__/    Hubble Relay:       disabled
    \__/       ClusterMesh:        disabled

Deployment             cilium-operator    Desired: 1, Ready: 1/1, Available: 1/1
DaemonSet              cilium-envoy       Desired: 2, Ready: 2/2, Available: 2/2
DaemonSet              cilium             Desired: 2, Ready: 2/2, Available: 2/2
Containers:            cilium-envoy       Running: 2
                       cilium             Running: 2
                       cilium-operator    Running: 1
Cluster Pods:          3/3 managed by Cilium
Helm chart version:    1.16.0-dev
Image versions         cilium             localhost:5000/cilium/cilium-dev:local: 2
                       cilium-operator    localhost:5000/cilium/operator-generic:local: 1
                       cilium-envoy       quay.io/cilium/cilium-envoy:v1.28.1-0a4c2d1a90a7e13116bed4b0c1d4aacaf0e49686@sha256:9c45b847f0d6689b537000257dc26a1db3799fd40cb2d430397fd0aec375a562: 2
```

**Comment by fujitatomoya** [comment]
> Edit: We can improve the experience further by pre-install required tools (e.g. kind, kubectl, and cilium-cli), however, this can be done in separate PR.

so true, this happens every time recreating the devcontainer, which is likely to happen. i can consider this with another issue.

**Review by qmonnet — APPROVED** [review]
Looks OK as far as I can tell

**Pull request merged**